## Intermediate Data Science

#### University of Redlands - DATA 201
#### Prof: Joanna Bieri [joanna_bieri@redlands.edu](mailto:joanna_bieri@redlands.edu)
#### [Class Website: data201.joannabieri.com](https://joannabieri.com/data201_intermediate.html)

## Computer Set Up

In [6]:
!python --version

Python 3.12.2


In [8]:
!conda --version

conda 24.11.0


In [9]:
!git --version

git version 2.39.5 (Apple Git-154)


**Clone the Repo for our Class**

<a href="https://github.com/Redlands-DATA201/FALL25" target="_blank">Redlands-DATA201/FALL2025</a>

In [13]:
# # Depending on your setup you might need to install modules
# !conda install -y numpy
# !conda install -y pandas
# !conda install -y matplotlib
# !conda install -y plotly

In [15]:
# Some basic package imports
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.defaule = 'colab'

# Review of DATA 101

For our review we are going to jump in and do some Exploratory Data Analysis (EDA) on a data set that we have seen before. This time you are just given a .csv file and your goal is to analyze it - answering these questions.

**I expect this to feel impossible at first!!!** but if you all work together, share code, and look things up as needed, I know you can do it!

#### This was a Homework Assignment in Data 101 - but see how much you can figure out on your own!

* Answer the following questions using reproducible Python code.
    - What does it mean to be reproducible? This means someone else at any time in the future can run and understand your code almost like reading a blog post.
* For each question, state your answer in a sentence, e.g. "In this sample, the first three common names of purchasers are ...".
* Note that the answers to all questions are within the context of this particular sample of sales, i.e. you shouldn't make inferences about the population of all Lego sales based on this sample.

0.  Read in the .csv file using Pandas and display it

1.  Describe what you see in the data set (variables, observations, etc)

2.  What are the three most common first names of purchasers?

3.  What are the three most common themes of Lego sets purchased?

4.  Among the most common theme of Lego sets purchased, what is the most common subtheme?

5.  Create data frames for each of the ages in the following categories: "18 and under", "19 - 25", "26 - 35", "36 - 50", "51 and over". HINT - use masks

6.  Which age group has purchased the highest number of Lego sets.

7.  Which age group has spent the most money on Legos?

8.  Which Lego theme has made the most money for Lego?

9.  Which area code has spent the most money on Legos? In the US the area code is the first 3 digits of a phone number. Then using a for loop calculate the average money spent per customer for each area code.

10.  Come up with a question you want to answer using these data, and write it down. Then, create a data visualization that answers the question, and explain how your visualization answers the question.

**This exersize comes from: https://datasciencebox.org/course-materials/hw-instructions/hw-05/hw-05-legos**

In [43]:
file_name = 'data/lego-sales.csv'
DF = pd.read_csv(file_name)
DF

first_name  last_name  age  phone_number  set_id   number  \
0     Kimberly  Beckstead   24  216-555-2549   24701    76062   
1         Neel     Garvin   35  819-555-3189   25626    70595   
2         Neel     Garvin   35  819-555-3189   24665    21031   
3      Chelsea   Bouchard   41           NaN   24695    31048   
4      Chelsea   Bouchard   41           NaN   25626    70595   
..         ...        ...  ...           ...     ...      ...   
615     Talise   Nieukirk   16  801-555-2343   24902    41556   
616    Spencer     Morgan   28  784-555-3455   26041    41580   
617    Spencer     Morgan   28  784-555-3455   26060  5005051   
618     Amelia    Hageman   40  336-555-1950   24702    76063   
619     Amelia    Hageman   40  336-555-1950   24720    10830   

                      theme              subtheme  year  \
0    DC Comics Super Heroes         Mighty Micros  2018   
1                   Ninjago  Rise of the Villains  2018   
2              Architecture                   NaN  2018   
3                   Creator                   NaN  2018   
4                   Ninjago  Rise of the Villains  2018   
..                      ...                   ...   ...   
615                  Mixels              Series 7  2018   
616                  Mixels              Series 9  2018   
617                    Gear         Digital Media  2018   
618  DC Comics Super Heroes         Mighty Micros  2018   
619                   Duplo                   NaN  2018   

                                       name  pieces  us_price  \
0                            Robin vs. Bane    77.0      9.99   
1                      Ultra Stealth Raider  1093.0    119.99   
2                              Burj Khalifa   333.0     39.99   
3                            Lakeside Lodge   368.0     29.99   
4                      Ultra Stealth Raider  1093.0    119.99   
..                                      ...     ...       ...   
615                                  Tiketz    62.0      4.99   
616                                    Myke    63.0      4.99   
617  Friends of Heartlake City Girlz 4 Life     NaN     19.99   
618              The Flash vs. Captain Cold    88.0      9.99   
619                           Minnie's Café    27.0     19.99   

                                             image_url  quantity  
0    http://images.brickset.com/sets/images/76062-1...         1  
1    http://images.brickset.com/sets/images/70595-1...         1  
2    http://images.brickset.com/sets/images/21031-1...         1  
3    http://images.brickset.com/sets/images/31048-1...         1  
4    http://images.brickset.com/sets/images/70595-1...         1  
..                                                 ...       ...  
615  http://images.brickset.com/sets/images/41556-1...         2  
616                                                NaN         2  
617                                                NaN         1  
618  http://images.brickset.com/sets/images/76063-1...         2  
619  http://images.brickset.com/sets/images/10830-1...         4  

[620 rows x 14 columns]

In [41]:
DF.describe()

age        set_id    year       pieces    us_price    quantity
count  620.000000    620.000000   620.0   551.000000  620.000000  620.000000
mean    34.356452  25124.982258  2018.0   254.206897   29.041613    1.437097
std     11.276537    506.760720     0.0   357.738804   34.630623    0.712849
min     16.000000  24548.000000  2018.0    13.000000    3.990000    1.000000
25%     25.000000  24724.750000  2018.0    70.000000    9.990000    1.000000
50%     33.000000  24804.500000  2018.0   114.000000   19.990000    1.000000
75%     41.000000  25640.250000  2018.0   313.000000   29.990000    2.000000
max     68.000000  26060.000000  2018.0  4634.000000  349.990000    5.000000

In [50]:
print(DF.dtypes)
print(DF.shape)

first_name       object
last_name        object
age               int64
phone_number     object
set_id            int64
number           object
theme            object
subtheme         object
year              int64
name             object
pieces          float64
us_price        float64
image_url        object
quantity          int64
dtype: object
(620, 14)


Looking at the data frame above we can see that there is 14 variables, and 620 observations.